tableaudocumentapi/datasource.py

```diff
@@ -258,3 +259,38 @@ def _get_metadata_objects(self):
     def _get_column_objects(self):
         return [_column_object_from_column_xml(self._datasourceTree, xml)
                 for xml in self._datasourceTree.findall('.//column')]
+
+    ###########
+    # folders
+    ###########
+    @property
+    def folders(self):
+        if not self._folders:
+            self._refresh_folders()
+        return self._folders
+
+    def _refresh_folders(self):
+        folders = Folder.all_folders_from_datasource(self)
+        self._folders = {f.name: f for f in folders}
+
+    def add_folder(self, name, role):
+        """ Adds a new, empty folder to the datasource and returns it.
+
+        Will fail if another folder of the same name already exists,
+        or if no valid role (dimensions or measures) is provided.
+        """
+
+        if name in self.folders.keys():
+            raise ValueError('Folder names must be unique')
+
+        folder = Folder.from_name_and_role(self, name, role)
+        # find the index of the last column to insert the folder
```

Comment:
If you create/export a datasource in Tableau Desktop, the resulting .tds looks like:

```
<datasource>
  ....
  <column> ....</column>
  <column> ....</column>
  <folder .../>
  <folder .../>
  ...
</datasource>
```

The <folder> tags are always below the <column> tags, despite both being children of datasource. For the XML parser, this doesn't necessarily make a difference. But for (1) parity with the way Tableau Desktop creates TDS files and (2) better readability for humans, I've decided to emulate this behaviour. To this end, I look for the index of the last <column> tag and insert the folder after it. 

I hope that this explanation is somehow understandable. Does this answer your question?